File: packages/server/internal/game/map/map.go
package gamemap

import (
	"server/internal/game/block"
	"strconv"
)

type Size struct {
	Width  uint16 `json:"width"`
	Height uint16 `json:"height"`
}

type Pos struct {
	X uint16
	Y uint16
}

type Info struct {
	Id   string `json:"id"`
	Name string `json:"name"`
	Desc string `json:"desc"`
}

type Sight [][]bool // 2D slice for visibility, true if visible

type Blocks [][]block.Block // 2D slice for blocks

type Map interface {
	IsEmpty() bool
	Block(pos Pos) (block.Block, error)
	Blocks() Blocks
	SetBlock(pos Pos, b block.Block) error
	SetBlocks(blocks Blocks) error // Set all blocks at once
	Size() Size
	Info() Info

	RoundStart(roundNum uint16)
	RoundEnd(roundNum uint16)

	Fog(owner []block.Owner, sight Sight) error
}

func (b Blocks) String() string {
	// return a debug string representation of the blocks
	// TODO
	return "Blocks(...)"
}

func (s Size) String() string {
	return strconv.Itoa(int(s.Width)) + "x" + strconv.Itoa(int(s.Height))
}

func (p Pos) String() string {
	return "Pos(" + strconv.Itoa(int(p.X)) + "," + strconv.Itoa(int(p.Y)) + ")"
}

func (s Size) IsPosValid(p Pos) bool {
	return p.X >= 1 && p.X <= s.Width && p.Y >= 1 && p.Y <= s.Height
}

func (i Info) String() string {
	return "Info(#" + i.Id + ", " + i.Name + ", Desc: " + i.Desc + ")"
}

File: packages/shared-types/src/map.ts
import { Block, Owner } from './block';

// Size interface
export interface Size {
  width: number; // uint16 in Go
  height: number; // uint16 in Go
}

// Pos interface
export interface Pos {
  x: number; // uint16 in Go
  y: number; // uint16 in Go
}

// Info interface
export interface Info {
  id: string;
  name: string;
  desc: string;
}

// Type aliases for clarity
export type Sight = boolean[][]; // 2D array for visibility, true if visible
export type Blocks = Block[][]; // 2D array for blocks

// Map interface
export interface Map {
  isEmpty(): boolean;
  block(pos: Pos): Block | null; // Returns null instead of error in TS
  blocks(): Blocks;
  setBlock(pos: Pos, b: Block): void; // No error return in TS
  setBlocks(blocks: Blocks): void; // Set all blocks at once
  size(): Size;
  info(): Info;
  
  roundStart(roundNum: number): void;
  roundEnd(roundNum: number): void;
  
  fog(owner: Owner[], sight: Sight): void; // No error return in TS
}

// Helper functions
export function sizeToString(s: Size): string {
  return `${s.width}x${s.height}`;
}

export function posToString(p: Pos): string {
  return `Pos(${p.x},${p.y})`;
}

export function isPosValid(s: Size, p: Pos): boolean {
  return p.x >= 1 && p.x <= s.width && p.y >= 1 && p.y <= s.height;
}

export function infoToString(i: Info): string {
  return `Info(#${i.id}, ${i.name}, Desc: ${i.desc})`;
}
